Implementing acceptance criteria Ticket 6352

Adam-Szw opened the pull request:
Changed code to accomodate variable amount of columns in lookup file. Implemented functions for creating or reading parameters dynamically. Added multiple checks to prevent IOC from crashing in unexpected scenarios like editing file mid-run or incorrect formatting of file

[Ticket 6352](https://github.com/ISISComputingGroup/IBEX/issues/6352)

Commit:
Changed code to accomodate variable amount of columns in lookup file. Implemented functions for creating or reading parameters dynamically. Added multiple checks to prevent IOC from crashing in unexpected scenarios like editing file mid-run or incorrect formatting of file

Adam-Szw commented:
Tests written, PR: https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/476

FreddieAkeroyd committed:
VS2010 fixes

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -187,16 +177,12 @@ asynStatus ReadASCII::writeInt32(asynUser *pasynUser, epicsInt32 value)
+                    setParamTableValue(it->first, it->second, value);
                 }
+                setParamTableValue(config_setpoint_value.first, config_setpoint_value.second, value);
             }
```
I think we have lost the `!quietOnSetPoint` printing?

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -326,24 +312,27 @@ asynStatus ReadASCII::readFloat64Array(asynUser *pasynUser, epicsFloat64 *value,
+            try
+            {
+                if (function == it->second) {
+                    memcpy(value, pv_data_column_reference[it->first]->data(), ncopy * sizeof(epicsFloat64));
```
`ncopy` is no longer assigned a value before it is used
```
C:\Instrument\Apps\EPICS\support\ReadASCII\master\ReadASCIIApp\src\ReadASCII.cpp(323) : warning C4700: uninitialized local variable 'ncopy' used
```

FreddieAkeroyd replied:
I think you could also use
```
std::copy_n(pv_data_column_reference[it->first]->cbegin(), ncopy, value);
```

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -326,24 +312,27 @@ asynStatus ReadASCII::readFloat64Array(asynUser *pasynUser, epicsFloat64 *value,
+            std::cerr << "ReadASCII: WARNING - parameter name: " << it->first << " doesn't have associated table column" << std::endl;
+        }
     }
-    *nIn = ncopy;
```
You need to set `*nIn` - this tells epics how many values you returned. On success `*nIn = ncopy`, on error you should set `*nIn = 0`

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -499,28 +487,26 @@ int ReadASCII::getSPInd (double SP)
+        epicsFloat64 ret = setParamTableValue(it->first, it->second, index);
+        prnt += it->first + ":" + std::to_string(static_cast<long double>(ret)) + ", ";  // long double cast needed for VS2010
+    }
+    std::cerr << prnt + "\n";
```
Using an `ostringstream` rather than string concatenation might be better - write to this as is you were writing to `cerr` and at the end use   `std::cerr << oss.str() << std::endl`

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.h:
```diff
@@ -43,37 +38,62 @@ class ReadASCII : public asynPortDriver
-    int P_I; //float
-    int P_D; //float
-    int P_MaxHeat; //float
+	int* lastParam;
```
The version of asyn we are currently using does not need to know the number of parameters, so the `lastParam` logic can be removed and also the calculation of `NUM_READASCII_PARAMS` and passing this to `asynPortDriver` in the `ReadASCII` constructor

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -576,61 +562,269 @@ void ReadASCII::readFilePoll(void)
+    }
+    out.push_back(line);
+    return out;
+}
```
As you are splitting on a single character, you can use `getline`
```
std::istringstream iss(line);
std::string str;
while(std::getline(iss, str, ' ')) { out.push_back(str); }
```

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -93,6 +74,15 @@ ReadASCII::ReadASCII(const char *portName, const char *searchDir, const int step
+
     // Do no initialise P_SPOut here because it will be sent to the eurotherm
 
     lastModified = 0;
```
The line below this says
```
fileBad = true; //Set so that program doesn't attempt to read file before base dir is set
```
however `createParams` reads the file. Does the comment just need removing?

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -93,6 +74,15 @@ ReadASCII::ReadASCII(const char *portName, const char *searchDir, const int step
     setIntegerParam(P_RampOn, 0);
     setIntegerParam(P_LookUpOn, 0);
 
+    char localDir[DIR_LENGTH], dir[DIR_LENGTH];
```
Does `dir` need to have size `2*DIR_LENGTH + 1` ? It gets `localDir` and `/` concatenated onto it

FreddieAkeroyd commented on ReadASCIIApp/src/ReadASCII.cpp:
```diff
@@ -576,61 +562,269 @@ void ReadASCII::readFilePoll(void)
 
-asynStatus ReadASCII::readFile(const char *dir)
+asynStatus ReadASCII::readFile(const char *dir, bool retry)
 {
```
The function no longer sets `fileBad = true` on error - just checking that it was no longer required

FreddieAkeroyd commented:
Thanks for your work on generalising this - a lot of comments above, but mostly minor. A couple of more general points:
- the original code used the word `index` to mean the row of the table being used, however with the generalisation of column names there are arrays with the name `index` in them that refer to asyn parameter indexes. This could be potentially confusing to a reader e.g. ` updateControlValues` takes an index parameter that is actually a `tableRow` no an asyn parameter index.
- it would be nice to improve the interplay of `fileBad` and `isModified` logic in `readFilePoll` - maybe  `isModified` should be setting `fileBad` ?

Commit:
Code reworked to make it easier to read and debug. Asyn parameters can be set dynamically. Implemented feedback from last review. Added even more generalisation.

FreddieAkeroyd committed:
Various fixes/additions

Add some googletest tests
Checked for array prefix using compare rather than find
Fix some information prints
Other code tidy ups
